lib/galaxy/webapps/galaxy/services/visualizations.py

```diff
@@ -60,3 +97,284 @@ def index(
+        """
+        revision_dict = {
+            "model_class": "VisualizationRevision",
+            "id": revision.id,
+            "visualization_id": revision.visualization.id,
+            "title": revision.title,
+            "dbkey": revision.dbkey,
+            "config": revision.config,
+        }
+        return VisualizationRevisionResponse(**revision_dict)
+
+    def _add_visualization_revision(
+        self,
+        trans: ProvidesUserContext,
+        visualization: Visualization,
+        config: Optional[Union[dict, bytes]],
+        title: Optional[str],
+        dbkey: Optional[str],
+    ) -> VisualizationRevision:
+        """
+        Adds a new `VisualizationRevision` to the given `visualization` with
+        the given parameters and set its parent visualization's `latest_revision`
+        to the new revision.
+        """
+        # precondition: only add new revision on owned vis's
+        # TODO:?? should we default title, dbkey, config? to which: visualization or latest_revision?
+        revision = VisualizationRevision(visualization=visualization, title=title, dbkey=dbkey, config=config)
+
+        visualization.latest_revision = revision
+        # TODO:?? does this automatically add revision to visualzation.revisions?
+        trans.sa_session.add(revision)
+        with transaction(trans.sa_session):
+            trans.sa_session.commit()
+        return revision
+
+    def _create_visualization(
+        self,
+        trans: ProvidesUserContext,
+        type: str,
+        title: Optional[str] = None,
+        dbkey: Optional[str] = None,
+        slug: Optional[str] = None,
+        annotation: Optional[str] = None,
+        save: Optional[bool] = True,
+    ) -> Visualization:
+        """Create visualization but not first revision. Returns Visualization object."""
+        user = trans.get_user()
+
+        # Error checking.
+        title_err = slug_err = ""
+        if not title:
+            title_err = "visualization name is required"
+        elif slug and not is_valid_slug(slug):
+            slug_err = "visualization identifier must consist of only lowercase letters, numbers, and the '-' character"
+        elif slug and slug_exists(trans.sa_session, Visualization, user, slug, ignore_deleted=True):
+            slug_err = "visualization identifier must be unique"
+
+        if title_err or slug_err:
+            # TODO: handle this error structure better
+            val_err = str(title_err or slug_err)
+            raise exceptions.RequestParameterMissingException(val_err)
+
+        # Create visualization
+        visualization = Visualization(user=user, title=title, dbkey=dbkey, type=type)
+        if slug:
+            visualization.slug = slug
+        else:
+            slug_builder = SlugBuilder()
+            slug_builder.create_item_slug(trans.sa_session, visualization)
+        if annotation:
+            annotation = sanitize_html(annotation)
```

Comment:
I don't think we should or need to do this, but also you're already doing this in the incoming pydantic model.